File: src/main/java/ru/frostman/scalable/tpc/TpcExample.java
package ru.frostman.scalable.tpc;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;

/**
 * @author slukjanov aka Frostman
 */
public class TpcExample {

    public static void main(String[] args) {

    }

    public void dispatch(int port) {
        ServerSocket serverSocket;
        try {
            serverSocket = new ServerSocket(port);
        } catch (IOException e) {
            System.err.println("Can't open port " + port);
            e.printStackTrace();
            System.exit(-1);
            return;
        }

        while (true) {
            Socket socket = null;
            try {
                socket = serverSocket.accept();
            } catch (IOException e) {
                System.err.println("Can't accept new connection");
                e.printStackTrace();
            }

            if (socket == null) {
                continue;
            }

            new Worker(socket).start();
        }
    }

    public final class Worker implements Runnable {
        private final Thread thread = new Thread(this);
        private final Socket socket;

        public Worker(Socket socket) {
            this.socket = socket;
        }

        public void start() {
            thread.start();
        }

        public void run() {
            DataInputStream in = null;
            DataOutputStream out = null;
            try {
                in = new DataInputStream(socket.getInputStream());
                out = new DataOutputStream(socket.getOutputStream());

                out.writeInt(in.readInt() + 1);

            } catch (IOException e) {
                System.err.println("IO problems");
                e.printStackTrace();
            } finally {
                try {
                    if (in != null) {
                        in.close();
                    }
                    if (out != null) {
                        out.close();
                    }
                    socket.close();
                } catch (Exception e) {
                    // no operations
                }
            }
        }
    }

}

File: src/tpc/TpcExample.kt
package tpc

/**
 * @author slukjanov aka Frostman
 */

import java.io.PrintWriter
import java.util.concurrent.atomic.AtomicInteger
import java.net.ServerSocket
import java.net.Socket
import java.io.DataInputStream
import java.io.DataOutputStream
import java.util.concurrent.CountDownLatch
import java.net.SocketException

fun main(args : Array<String>) = TpcExample(100).demo()

public class TpcExample(val clients : Int) {
    private val nextWorkerId = AtomicInteger(1)
    private val output = PrintWriter(System.out)
    private val notProcessed = CountDownLatch(clients)

    fun demo() {
        val start = System.currentTimeMillis()
        log("Starting tpc demo with $clients clients")

        val server = TpcServer(8000)
        server.start()
        Thread.sleep(100)
        for (i in 1..clients) TpcClient(8000, i).start()

        notProcessed.await()
        server.server.close()
        log("All clients processed, demo takes about ${System.currentTimeMillis() - start}ms.")

        output.close()
    }

    public class TpcServer(val port : Int) : Thread("tpc demo server") {
        val server = ServerSocket(port, 100000)

        override fun run() {
            while (notProcessed.count > 0) {
                try {
                    val socket = server.accept().sure()
                    Worker(socket).start()
                } catch (e : SocketException) {
                    // no operations
                }
            }
        }
    }

    class Worker(val socket : Socket) : Thread("worker #${nextWorkerId.getAndIncrement()}") {
        override fun run() {
            Thread.sleep(100)
            log("new connection accepted by $name")
            val input = DataInputStream(socket.getInputStream())
            val output = DataOutputStream(socket.getOutputStream())
            output.writeInt(input.readInt() + 1) // increment received Int
            log("connection successfully processed by $name")
        }
    }

    public class TpcClient(val serverPort : Int, val cid : Int) : Thread("client #$cid") {
        override fun run() {
            log("$name connecting to localhost:$serverPort")
            val socket = Socket("127.0.0.1", serverPort)
            val input = DataInputStream(socket.getInputStream())
            val output = DataOutputStream(socket.getOutputStream())
            output.writeInt(100)
            if (input.readInt() != 101) {
                throw IllegalStateException()
            }
            socket.close()
            notProcessed.countDown()
            log("$name successfully proceed")
        }
    }

    fun log(str : String?) = output.println("$str")
}
